### ISSUE TITLE
SqlPersistenceSession throws in endpoint without sagas/outbox

### ISSUE THREAD

**Issue by DavidBoike** [Particular/NServiceBus.Persistence.Sql#270]
In an endpoint without Sagas or Outbox, there is no reason for the endpoint to need to open a DbConnection for its own persistence needs, so it doesn't. Then when the user attempts to call the following, the result is an exception:

```
context.SynchronizedStorageSession.SqlPersistenceSession()
```

Exception comes from [this line of code](https://github.com/Particular/NServiceBus.Persistence.Sql/blob/master/src/SqlPersistence/SqlPersistenceStorageSessionExtensions.cs#L31):

> Exception: The endpoint has not been configured to use SQL persistence.

This is "normal" behavior but is a usability bug. Try to move a bare handler from a busy endpoint with Sagas to an endpoint all by itself for performance reasons and it will suddenly blow up. It should be possible for SQL Persistence to lazily create the DbConnection according to its connection builder rules and then handle completing it within the Unit of Work if it turns out it got created.

This is most likely to occur when using SQL Persistence with SQL Transport because you can use the same connection string and get full message+persistence+business consistency with just a local database transaction, and without using Outbox. Probably also occurs when using SQL Persistence + MSMQ + DTC although I haven't tested it.

## Reproducing

1. Download the [](https://docs.particular.net/samples/sqltransport-sqlpersistence/) sample
2. Remove the `OrderLifecycleSaga` class in the Receiver, which is just one of many handlers for the `OrderSubmitted` event. This leaves Receiver having only handlers, no sagas.
3. Run, and press <kbd>Enter</kbd> in the Sender project to publish an event. Exception will occur in the Receiver.

## Workaround

1. In these handlers there should be no NServiceBus reason to use persistence, so you can create your own DbConnection using the same connection string.
2. If using any transport besides SQL Transport or MSMQ, enable the Outbox, which should probably be used to guarantee consistency between persistence and message transport anyway.

**Comment by DavidBoike** [comment]
We do have a warning about this in doco:

> The session (`context.SynchronizedStorageSession.SqlPersistenceSession()`) is only supported when the SQL persistence has been selected to be used for Sagas and/or the Outbox. See also Persistence Storage Types and Configuration order for persistence.

...but that doesn't provide any rationale or make much sense to the end user.

**Comment by mauroservienti** [comment]
We fixed the same issue in RavenDB a while ago, but it’s causing other headaches https://github.com/Particular/NServiceBus.RavenDB/issues/365

**Comment by SzymonPobiega** [comment]
We discussed this with Simon while working on SQL persistence. We decided to use different approach than NH (which provides the session always) and only provide it when sagas or outbox is enabled.

The rationale is that in absence of sagas/outbox users really should not depend on NSB for their data access because we can't guarantee anything more than he can. I think the exception message should provide better guidance.

**Comment by markgould** [comment]
I just ran into this after upgrading from 2.1.4 to 3.1.0. I agree that the exception should provide better guidance. This did work in 2.1.4, so it does appear to be a breaking change.

**Comment by SzymonPobiega** [comment]
After having discussed that with @weralabaj and @tmasternak we think we should change the SQL persistence so it behaves like NHibernate i.e. make the session always available, not only when sagas and/or outbox is enabled.

**Comment by SzymonPobiega** [comment]
I did some research and found out that NServiceBus code uses smart way of merging persistence registrations (`config.UsePersistence) using [this class](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Persistence/PersistenceStorageMerger.cs). The effect is that only a single persistence/concern combo gets *configured* in any given endpoint. The merge process uses *last registration wins* approach. By configured I mean the code in *supports* clause like this gets invoked:
```
Supports<StorageType.Outbox>(s =>
{
    EnableSession(s);
    s.EnableFeatureByDefault<SqlOutboxFeature>();
});
```

This does not mean, yet, that the `SqlOutboxFeature` is going to be activated -- it might not meet the prerequisites. 

Based on that we should enable the SQL synchronized storage session **whenever outbox or saga features are configured**. This essentially means that no other persistence technology claimed these concerns. It might happen that this endpoint has outbox and sagas disabled but we still should run the session because handlers might need it.

Using this approach we also allow for reasonable mixing of persistence concerns e.g. using SQL persistence for subscriptions and NHibernate for the rest. In this case SQL persistence does not claim the responsibility for outbox or saga concern so it won't run its session.

**Comment by weralabaj** [comment]
Clarified with @SzymonPobiega, the idea is:
- To revert the functionality to what was done in 2.1.4, i.e. save synchronized storage session even if neither saga nor outbox are enabled, so essentially remove this line added in 3.1.0 https://github.com/Particular/NServiceBus.Persistence.Sql/blob/develop/src/SqlPersistence/SynchronizedStorage/StorageSessionFeature.cs#L30.
- Add verification step and throw if we use different persistences per saga and outbox (we need both to be SqlP, so that the connection/session is used correctly).
   - Add similar check in NHibernate, if needed.

**Comment by SzymonPobiega** [comment]
Closed via #290
